### PR TITLE
fix high cpu

### PR DESCRIPTION
**Pull request by jinzhongjia**
Recently, my main development environment was changed to windows. I found that the event used by git-blame is cursormoved, and this event is triggered very frequently.

Git itself on windows runs slower, which causes the entire neovim to be blocked

I replaced cursormoved with cursorhold

### CODE REVIEW

**Comment by jinzhongjia** [comment]
for some users, they may like use cursormoved to get get an immediate response, we can implement an debounce function

**Comment by jinzhongjia** [comment]
this pr can solved there issue: 
- https://github.com/f-person/git-blame.nvim/issues/83
- https://github.com/f-person/git-blame.nvim/issues/119
- https://github.com/f-person/git-blame.nvim/issues/52

**Comment by jinzhongjia** [comment]
I decided to implement additional debounce functionality, although CursorHold is enough for me

**Comment by jinzhongjia** [comment]
This pr add three config options:
- `gitblame_schedule_event`
- `gitblame_clear_event`
- `gitblame_debounce_delay`
- 
Instructions have been added to the readme

**Comment by f-person** [comment]
Hi! Thanks for the PR! I'll check it in the near future! 
In the meantime, is the debounce functionality different from https://github.com/f-person/git-blame.nvim?tab=readme-ov-file#visual-delay-for-displaying-the-blame-info?

**Comment by jinzhongjia** [comment]
Oh, I didn't notice this before
I looked at the code and noticed that you also used `vim.defer_fn`, but why do you use `vim.loop.is_active` to detect whether `delay_timer` is valid?
As far as I know loop as luv's binding does not recognize deferred tasks created via `vim.defer_fn`

**Comment by jinzhongjia** [comment]
And I noticed that this logic is only valid for `schedule_show_info_display`

**Comment by jinzhongjia** [comment]
I implemented debounce, `schedule` and `clear virtual text` for both

**Comment by jinzhongjia** [comment]
But now it is obvious that the feature is duplicated, I will submit a new commit to merge them

**Comment by jinzhongjia** [comment]
ok, `vim.loop.is_active` can detect the content returned by `vim.defer_fn`

**Comment by jinzhongjia** [comment]
Now, I think every thing is ok, current code only enhances the original functions, adds event configurations for `schedule_show_info_display` and `clear_virtual_text`, and debounce time is 250 milliseconds by default.

File: README.md [hunk 1]
```diff
@@ -203,7 +204,7 @@ The delay in milliseconds after which the blame info will be displayed.
 
 Note that this doesn't affect the performance of the plugin.
 
-Default: `0`
+Default: `250`
```
Comment (bossley9) [inline README.md:207]: I'm not sure that we should change this default value as it is a breaking change. Maybe include a note instead in the performance section to change this to `250` for better results?

File: lua/gitblame/init.lua [hunk 2]
```diff
@@ -627,12 +619,84 @@ local function clear_all_extmarks()
...
+        timer = vim.defer_fn(function()
+            func(unpack(args))
+            timer = nil
+        end, delay)
+    end
+end
+
+-- this function is uesed to verify the config info for debounce
+--- @return boolean
+local verify_debounce_info = function()
```
Comment (bossley9) [inline lua/gitblame/init.lua:643]: Rather than implementing a hard fail if the values don't match the expected values, it might be better to add this in another pull request since we don't really do any value checking on any of the configuration values and I think it would be good to separately add it for all configuration values. Thoughts?

Comment (bossley9) [reply lua/gitblame/init.lua:643]: We could do something like `validate_configuration_options` which validates all the values at once

File: lua/gitblame/init.lua [hunk 3]
```diff
@@ -627,12 +619,84 @@ local function clear_all_extmarks()
...
+            timer:stop()
+            timer = nil
+        end
+
+        timer = vim.defer_fn(function()
+            func(unpack(args))
+            timer = nil
+        end, delay)
+    end
+end
```
Comment (bossley9) [inline lua/gitblame/init.lua:639]: Nitpick: we should move this to `utils`

**Review by bossley9 — COMMENTED** [review]
This is great! I left a few comments mostly on pull request polishing.

File: README.md (hunk 1 above)
Comment (jinzhongjia) [reply README.md:207]: In fact, the reasons for the above issues are all caused by not turning on debounce by default. I think it should be prioritized to ensure that users use normal settings by default.

File: lua/gitblame/init.lua (hunk 2 above)
Comment (jinzhongjia) [reply lua/gitblame/init.lua:643]: I don't know all the configurations, so this is all I can do here

File: README.md [hunk 4]
```diff
@@ -221,7 +222,27 @@ Default: `v:null`
...
+
+If you want better performance (e.g. in particularly large projects) you can refer here
+
+`g:gitblame_schedule_event`
+
+the event used for schedule
+
+value: `CursorMoved`|`CursorHold` 
+
+default is `CursorMoved`, `CursorHold` will be better performance, but the response will be slightly slower
```
[the remark this answers is not in the document]
Comment (jinzhongjia) [reply README.md]: of course

File: README.md [hunk 5]
```diff
@@ -221,7 +222,27 @@ Default: `v:null`
...
+
+`g:gitblame_clear_event`
+
+the evene used for clear virtual text
+
+value: `CursorMovedI`|`CursorHoldI`
+
+default is `CursorMovedI`, `CursorHoldI` will be better performance, but the response will be slightly slower
```
[the remark this answers is not in the document]
Comment (jinzhongjia) [reply README.md:241]: yes

File: README.md (hunk 1 above)
Comment (bossley9) [reply README.md:207]: Well, the only problem is that it changes the default behavior. I'm ok with this change but we just need to be aware of this

File: lua/gitblame/init.lua (hunk 2 above)
Comment (bossley9) [reply lua/gitblame/init.lua:624]: Ok, I can look into adding something like this after this pull request

File: README.md (hunk 1 above)
Comment (jinzhongjia) [reply README.md:207]: I think just setting the commit message to [break change] after squashing this PR is ok

File: README.md [hunk 6]
```diff
@@ -221,7 +222,25 @@ Default: `v:null`
...
+
+options: `CursorMoved`|`CursorHold` 
+
+`g:gitblame_clear_event`
+
+the evene used for clear virtual text
+
+value: `CursorMovedI`|`CursorHoldI`
+
+default is `CursorMovedI`, `CursorHoldI` will be better performance, but the response will be slightly slower
```
Comment (bossley9) [inline README.md:241]: Let's do the same here

```suggestion
`g:gitblame_clear_event` is used for clearing virtual text. See [CursorMovedI](https://neovim.io/doc/user/autocmd.html#CursorMovedI) and [CursorHoldI](https://neovim.io/doc/user/autocmd.html#CursorHoldI).

Default: `CursorMovedI`

options: `CursorMovedI`|`CursorHoldI`
```

**Comment by bossley9** [comment]
Other than that, I'm ok with this as long as @f-person is ok with it! :rocket:

**Comment by jinzhongjia** [comment]
ok, thx

**Comment by bossley9** [comment]
@f-person have you had any time to take a look at this?

File: README.md (hunk 1 above)
Comment (f-person) [reply README.md:207]: this is something i was also not sure about when i initially checked the pr, but didn't have the time to reply. i do think it could indeed be the right thing to do, but still unsure, unfortunately.

**Review by f-person — APPROVED** [review]
thank you, this is really great! i approve this pr, but left a comment; let's discuss that matter a bit further and maybe merge it or decide to do something else. looking forward from hearing back from you! thanks! <3

File: README.md (hunk 1 above)
Comment (jinzhongjia) [reply README.md:207]: > this is something i was also not sure about when i initially checked the pr, but didn't have the time to reply. i do think it could indeed be the right thing to do, but still unsure, unfortunately.

We can squash the entire PR and set the commit message is  `break change`, so that people who update the plugin can notice this change.

Comment (jinzhongjia) [reply README.md:207]: I have been using the pr branch to work after the submit. It currently performs well on windows and linux, and there has been no lag.

Comment (jinzhongjia) [reply README.md:207]: This is also the solution for other plugins to modify the default settings. In fact, this setting change is insensitive to most users.
It only affects those users who have modified `delay`. I don’t think anyone will move the cursor and look at the commit message of quick switching at the same time.

Comment (f-person) [reply README.md:207]: Yup, that makes sense :) 
I'd rather not use squash, though

**Comment by bossley9** [comment]
I linked the issues you mentioned since it is very likely this PR will fix them. The best way to test a big repository is to clone https://github.com/nixos/nixpkgs and try to open `pkgs/top-level/all-packages.nix` :smile:

**Review by bossley9 — APPROVED** [review]

**Comment by jinzhongjia** [comment]
I tried it briefly. At least for now, gitblame will not prevent me from scrolling, but when git gets the commit message in the background, there will still be a high CPU usage.

**Comment by jinzhongjia** [comment]
By the way, it is recommended to close all treesitter related plugins during testing, including itself, because the performance of treesitter is really poor.

**Comment by f-person** [comment]
Okay, let's merge this! Thank you very much for the effort on this, @jinzhongjia & @bossley9!! <3 
Ping me if you're ever in the Netherlands, and I'll get you a coffee or a beer :)

**Comment by f-person** [comment]
Maybe clearing the existing message should be done immediately on cursor move and not whenever we set the next one?

**Comment by jinzhongjia** [comment]
I think that's ok, let me create a new pr
